internal/vault/vault.go

```diff
@@ -187,7 +187,15 @@ func (v *Vault) newConfig() (*vault.Config, error) {
 
 	certs := v.issuer.GetSpec().Vault.CABundle
 	if len(certs) == 0 {
-		return cfg, nil
```

Comment:
Suggestion to clean up the logical code path a bit:

```
diff --git a/internal/vault/vault.go b/internal/vault/vault.go
index 01639cf48..25f9ee9cb 100644
--- a/internal/vault/vault.go
+++ b/internal/vault/vault.go
@@ -31,6 +31,7 @@ import (
 	corelisters "k8s.io/client-go/listers/core/v1"
 
 	v1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
 	"github.com/cert-manager/cert-manager/pkg/util/pki"
 )
 
@@ -185,23 +186,22 @@ func (v *Vault) newConfig() (*vault.Config, error) {
 	cfg := vault.DefaultConfig()
 	cfg.Address = v.issuer.GetSpec().Vault.Server
 
-	certs := v.issuer.GetSpec().Vault.CABundle
-	if len(certs) == 0 {
-		var err error
-		if v.issuer.GetSpec().Vault.CABundleSecretRef != nil {
-			certs, err = v.caBundleFromSecret()
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			return cfg, nil
-		}
+	caBundle, err := v.caBundle()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load vault CA bundle: %w", err)
+	}
+
+	// If no CA bundle was loaded, return early and don't modify the vault config
+	// further. This will cause the vault client to use the system root CA
+	// bundle.
+	if len(caBundle) == 0 {
+		return cfg, nil
 	}
 
 	caCertPool := x509.NewCertPool()
-	ok := caCertPool.AppendCertsFromPEM(certs)
+	ok := caCertPool.AppendCertsFromPEM(caBundle)
 	if !ok {
-		return nil, fmt.Errorf("error loading Vault CA bundle")
+		return nil, fmt.Errorf("no Vault CA bundles loaded, check bundle contents")
 	}
 
 	cfg.HttpClient.Transport.(*http.Transport).TLSClientConfig.RootCAs = caCertPool
@@ -209,8 +209,21 @@ func (v *Vault) newConfig() (*vault.Config, error) {
 	return cfg, nil
 }
 
-func (v *Vault) caBundleFromSecret() ([]byte, error) {
+// caBundle returns the CA bundle for the Vault server. Can be used in Vault
+// client configs to trust the connection to the Vault server. If no custom CA
+// bundle is configured, an empty byte slice is returned.
+// Assumes the in-line and Secret CA bundles are not both defined.
+// If the `key` of the Secret CA bundle is not defined, its value defaults to
+// `ca.crt`.
+func (v *Vault) caBundle() ([]byte, error) {
+	if len(v.issuer.GetSpec().Vault.CABundle) > 0 {
+		return v.issuer.GetSpec().Vault.CABundle, nil
+	}
+
 	ref := v.issuer.GetSpec().Vault.CABundleSecretRef
+	if ref == nil {
+		return nil, nil
+	}
 
 	secret, err := v.secretsLister.Secrets(v.namespace).Get(ref.Name)
 	if err != nil {
@@ -221,7 +234,7 @@ func (v *Vault) caBundleFromSecret() ([]byte, error) {
 	if ref.Key != "" {
 		key = ref.Key
 	} else {
-		key = "ca.crt"
+		key = cmmeta.TLSCAKey
 	}
 
 	certBytes, ok := secret.Data[key]
diff --git a/internal/vault/vault_test.go b/internal/vault/vault_test.go
index 54818bbc9..253c0f21f 100644
--- a/internal/vault/vault_test.go
+++ b/internal/vault/vault_test.go
@@ -28,7 +28,6 @@ import (
 	"errors"
 	"fmt"
 	"io"
-	clientcorev1 "k8s.io/client-go/listers/core/v1"
 	"net/http"
 	"strings"
 	"testing"
@@ -38,6 +37,7 @@ import (
 	"github.com/hashicorp/vault/sdk/helper/certutil"
 	"github.com/hashicorp/vault/sdk/helper/jsonutil"
 	corev1 "k8s.io/api/core/v1"
+	clientcorev1 "k8s.io/client-go/listers/core/v1"
 
 	vaultfake "github.com/cert-manager/cert-manager/internal/vault/fake"
 	cmapi "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
@@ -918,7 +918,7 @@ func TestNewConfig(t *testing.T) {
 					CABundle: []byte("a bad cert bundle"),
 				}),
 			),
-			expectedErr: errors.New("error loading Vault CA bundle"),
+			expectedErr: errors.New("no Vault CA bundles loaded, check bundle contents"),
 		},
 
 		"a good cert bundle should be added to the config": {
@@ -1028,7 +1028,7 @@ func TestNewConfig(t *testing.T) {
 					},
 				},
 				)),
-			expectedErr: errors.New("error loading Vault CA bundle"),
+			expectedErr: errors.New("no Vault CA bundles loaded, check bundle contents"),
 			fakeLister:  caBundleSecretRefFakeSecretLister("test-namespace", "bundle", "my-bundle.crt", "not a valid certificate"),
 		},
 	}

```